File: app/src/main/java/com/morsego/app/audio/MorseAudioSynthesizer.java
package com.morsego.app.audio;

import android.media.AudioAttributes;
import android.media.AudioFormat;
import android.media.AudioManager;
import android.media.AudioTrack;

import com.morsego.app.keyer.MorseTiming;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Ultra low-latency audio synthesizer in pure Java for CW sidetone.
 * Streams 16-bit PCM mono audio with smooth amplitude envelope to prevent clicks.
 */
public class MorseAudioSynthesizer {
    private static final int SAMPLE_RATE = 44100;
    private static final int BUFFER_SIZE = 1024;
    private static final int RAMP_SAMPLES = 176; // ~4ms cosine ramp at 44.1kHz

    private AudioTrack audioTrack;
    private final ExecutorService executorService;
    private final AtomicBoolean isRunning = new AtomicBoolean(false);
    private final AtomicBoolean isToneActive = new AtomicBoolean(false);

    private volatile float frequencyHz = 700.0f;
    private volatile float targetAmplitude = 0.0f;
    private float currentAmplitude = 0.0f;
    private double phase = 0.0;

    public MorseAudioSynthesizer() {
        this.executorService = Executors.newSingleThreadExecutor();
        initAudioTrack();
    }

    private void initAudioTrack() {
        int minBufferSize = AudioTrack.getMinBufferSize(
                SAMPLE_RATE,
                AudioFormat.CHANNEL_OUT_MONO,
                AudioFormat.ENCODING_PCM_16BIT
        );
        int bufferSize = Math.max(minBufferSize, BUFFER_SIZE * 4);

        AudioAttributes audioAttributes = new AudioAttributes.Builder()
                .setUsage(AudioAttributes.USAGE_MEDIA)
                .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                .build();

        AudioFormat audioFormat = new AudioFormat.Builder()
                .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                .setSampleRate(SAMPLE_RATE)
                .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                .build();

        audioTrack = new AudioTrack(
                audioAttributes,
                audioFormat,
                bufferSize,
                AudioTrack.MODE_STREAM,
                AudioManager.AUDIO_SESSION_ID_GENERATE
        );

        audioTrack.play();
        startAudioLoop();
    }

    private void startAudioLoop() {
        isRunning.set(true);
        executorService.execute(() -> {
            short[] buffer = new short[BUFFER_SIZE];
            float amplitudeStep = 1.0f / RAMP_SAMPLES;

            while (isRunning.get()) {
                if (!isToneActive.get() && currentAmplitude <= 0.0f) {
                    // Output silence when idle
                    for (int i = 0; i < BUFFER_SIZE; i++) {
                        buffer[i] = 0;
                    }
                    if (audioTrack != null && audioTrack.getPlayState() == AudioTrack.PLAYSTATE_PLAYING) {
                        audioTrack.write(buffer, 0, BUFFER_SIZE);
                    }
                    continue;
                }

                double twoPiFreq = 2.0 * Math.PI * frequencyHz / SAMPLE_RATE;

                for (int i = 0; i < BUFFER_SIZE; i++) {
                    if (currentAmplitude < targetAmplitude) {
                        currentAmplitude = Math.min(targetAmplitude, currentAmplitude + amplitudeStep);
                    } else if (currentAmplitude > targetAmplitude) {
                        currentAmplitude = Math.max(0.0f, currentAmplitude - amplitudeStep);
                    }

                    int sampleVal = (int) (Math.sin(phase) * currentAmplitude * 30000);
                    buffer[i] = (short) Math.max(Short.MIN_VALUE, Math.min(Short.MAX_VALUE, sampleVal));

                    phase += twoPiFreq;
                    if (phase > 2.0 * Math.PI) {
                        phase -= 2.0 * Math.PI;
                    }
                }

                if (audioTrack != null && audioTrack.getPlayState() == AudioTrack.PLAYSTATE_PLAYING) {
                    audioTrack.write(buffer, 0, BUFFER_SIZE);
                }
            }
        });
    }

    public void startTone() {
        isToneActive.set(true);
        targetAmplitude = 1.0f;
    }

    public void stopTone() {
        isToneActive.set(false);
        targetAmplitude = 0.0f;
    }

    public boolean isToneActive() {
        return isToneActive.get();
    }

    public void setFrequency(float freqHz) {
        this.frequencyHz = Math.max(300.0f, Math.min(1500.0f, freqHz));
    }

    public float getFrequency() {
        return frequencyHz;
    }

    /**
     * Plays a complete Morse sequence with PARIS timing asynchronously
     */
    public void playMorsePattern(String pattern, int wpm, Runnable onFinished) {
        playMorsePattern(pattern, wpm, 6, 13, onFinished);
    }

    public void playMorsePattern(String pattern, int wpm, int letterSpacingDits, int wordSpacingDits, Runnable onFinished) {
        Executors.newSingleThreadExecutor().execute(() -> {
            long dit = MorseTiming.ditDurationMs(wpm);
            long dah = MorseTiming.dahDurationMs(wpm);
            long intra = MorseTiming.intraCharSpaceMs(wpm);
            long inter = MorseTiming.interCharSpaceMs(wpm, letterSpacingDits);
            long word = MorseTiming.wordSpaceMs(wpm, wordSpacingDits);

            try {
                for (int i = 0; i < pattern.length(); i++) {
                    char c = pattern.charAt(i);
                    if (c == '.') {
                        startTone();
                        Thread.sleep(dit);
                        stopTone();
                        Thread.sleep(intra);
                    } else if (c == '-') {
                        startTone();
                        Thread.sleep(dah);
                        stopTone();
                        Thread.sleep(intra);
                    } else if (c == ' ') {
                        long extraPause = Math.max(0, inter - intra);
                        if (extraPause > 0) Thread.sleep(extraPause);
                    } else if (c == '/') {
                        long extraPause = Math.max(0, word - intra);
                        if (extraPause > 0) Thread.sleep(extraPause);
                    }
                }
            } catch (InterruptedException ignored) {
                stopTone();
            }

            if (onFinished != null) {
                onFinished.run();
            }
        });
    }

    public void release() {
        isRunning.set(false);
        stopTone();
        try {
            if (audioTrack != null) {
                audioTrack.stop();
                audioTrack.release();
            }
        } catch (Exception ignored) {}
        executorService.shutdownNow();
    }
}

File: app/src/main/java/com/morsego/app/audio/MorseAudioSynthesizer.kt
package com.morsego.app.audio

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

/**
 * Ultra low-latency audio synthesizer for Morse code CW sidetone.
 * Uses real-time AudioTrack PCM streaming with cosine-smoothed envelope to eliminate clicks.
 */
class MorseAudioSynthesizer(
    private var frequencyHz: Float = 700f
) {
    companion object {
        private const val SAMPLE_RATE = 44100
        private const val BUFFER_SIZE = 1024
        private const val RAMP_SAMPLES = 176 // ~4ms at 44.1kHz for anti-click envelope
    }

    private var audioTrack: AudioTrack? = null
    private var synthJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.Default)

    @Volatile
    private var isPlaying = false

    @Volatile
    private var targetAmplitude = 0f

    private var currentAmplitude = 0f
    private var phase = 0.0

    init {
        initAudioTrack()
    }

    private fun initAudioTrack() {
        val minBufferSize = AudioTrack.getMinBufferSize(
            SAMPLE_RATE,
            AudioFormat.CHANNEL_OUT_MONO,
            AudioFormat.ENCODING_PCM_16BIT
        )
        val bufferSize = maxOf(minBufferSize, BUFFER_SIZE * 4)

        audioTrack = AudioTrack.Builder()
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_MEDIA)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                    .build()
            )
            .setAudioFormat(
                AudioFormat.Builder()
                    .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                    .setSampleRate(SAMPLE_RATE)
                    .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                    .build()
            )
            .setBufferSizeInBytes(bufferSize)
            .setTransferMode(AudioTrack.MODE_STREAM)
            .build()

        audioTrack?.play()
        startAudioLoop()
    }

    private fun startAudioLoop() {
        synthJob?.cancel()
        synthJob = scope.launch {
            val shortBuffer = ShortArray(BUFFER_SIZE)
            val amplitudeStep = 1.0f / RAMP_SAMPLES

            while (isActive) {
                if (!isPlaying && currentAmplitude <= 0f) {
                    // Output silence when idle with minimal CPU
                    shortBuffer.fill(0)
                    audioTrack?.write(shortBuffer, 0, BUFFER_SIZE)
                    continue
                }

                val twoPiFreq = 2.0 * PI * frequencyHz / SAMPLE_RATE

                for (i in 0 until BUFFER_SIZE) {
                    // Smooth amplitude ramp to prevent pops/clicks
                    if (currentAmplitude < targetAmplitude) {
                        currentAmplitude = minOf(targetAmplitude, currentAmplitude + amplitudeStep)
                    } else if (currentAmplitude > targetAmplitude) {
                        currentAmplitude = maxOf(0f, currentAmplitude - amplitudeStep)
                    }

                    val sampleValue = (sin(phase) * currentAmplitude * 30000).toInt()
                    shortBuffer[i] = sampleValue.coerceIn(Short.MIN_VALUE.toInt(), Short.MAX_VALUE.toInt()).toShort()

                    phase += twoPiFreq
                    if (phase > 2.0 * PI) {
                        phase -= 2.0 * PI
                    }
                }

                audioTrack?.write(shortBuffer, 0, BUFFER_SIZE)
            }
        }
    }

    fun startTone() {
        isPlaying = true
        targetAmplitude = 1.0f
    }

    fun stopTone() {
        isPlaying = false
        targetAmplitude = 0.0f
    }

    fun setFrequency(newFreq: Float) {
        frequencyHz = newFreq.coerceIn(300f, 1500f)
    }

    fun getFrequency(): Float = frequencyHz

    suspend fun playToneDuration(durationMs: Long) {
        startTone()
        kotlinx.coroutines.delay(durationMs)
        stopTone()
    }

    fun release() {
        synthJob?.cancel()
        isPlaying = false
        targetAmplitude = 0f
        try {
            audioTrack?.stop()
            audioTrack?.release()
        } catch (_: Exception) {}
        audioTrack = null
    }
}
